stdlib/public/runtime/Paths.cpp

```diff
@@ -0,0 +1,276 @@
+
+  // Scan backwards until we find a path separator
+  const char *ptr = runtimePath + runtimePathLen;
+  while (ptr > runtimePath && !_swift_isPathSep(*--ptr));
+
+  // Remove the "lib" directory if it's present
+  if (ptr - runtimePath >= 4
+      && _swift_isPathSep(ptr[-4])
+      && std::strncmp(ptr - 3, "lib", 3) == 0) {
+    ptr -= 4;
+  }
+
+  // If the result is empty, return "./" or ".\\"
+  if (ptr == runtimePath) {
+    return "." PATHSEP_STR;
+  }
+
+  // Duplicate the string up to and including ptr
+  size_t len = ptr - runtimePath + 1;
+  char *thePath = (char *)malloc(len + 1);
+  std::memcpy(thePath, runtimePath, len);
+  thePath[len] = 0;
+
+  return thePath;
+}
+
+void
+_swift_initRootPath(void *)
+{
+  // SWIFT_ROOT overrides the path returned by this function
+  const char *swiftRoot = swift::runtime::environment::SWIFT_ROOT();
+  if (swiftRoot && *swiftRoot) {
+    size_t len = std::strlen(swiftRoot);
+
+    // Ensure that there's a trailing slash
+    if (_swift_isPathSep(swiftRoot[len - 1])) {
+      rootPath = swiftRoot;
+    } else {
+      char *thePath = (char *)malloc(len + 2);
+      std::memcpy(thePath, swiftRoot, len);
+      thePath[len] = PATHSEP_CHR;
+      thePath[len + 1] = 0;
+
+      rootPath = thePath;
+    }
+  } else {
+    rootPath = _swift_getDefaultRootPath();
+  }
+}
+
+}
+
+SWIFT_RUNTIME_EXPORT
+const char *
+swift_getAuxiliaryExecutablePath(const char *name)
+{
+  const char *rootPath = swift_getRootPath();
+
+  // Form <rootPath>/libexec/
+  size_t rootPathLen = std::strlen(rootPath);
+  const char *libexecStr = PATHSEP_STR "libexec" PATHSEP_STR;
+  size_t libexecLen = std::strlen(libexecStr);
+  char *libexecPath = (char *)malloc(rootPathLen + libexecLen + 1);
+  std::memcpy(libexecPath, rootPath, rootPathLen);
+  std::memcpy(libexecPath + rootPathLen, libexecStr, libexecLen + 1);
+
+  // If libexec exists, look there
+  if (_swift_exists(libexecPath)) {
+    const char *result = _swift_getAuxExePathIn(libexecPath, name);
+
+    free(libexecPath);
+
+    return result;
+  }
+
+  free(libexecPath);
+
+  // Otherwise, look in the root itself
+  return _swift_getAuxExePathIn(rootPath, name);
+}
+
+namespace {
+
+const char *
+_swift_getAuxExePathIn(const char *path, const char *name)
+{
+  size_t pathLen = std::strlen(path);
+  size_t nameLen = std::strlen(name);
+  size_t extLen = 0;
+
+#ifdef _WIN32
+  if (nameLen > 4 && strcmp(name + nameLen - 4, ".exe") != 0)
+    extLen = 4;
+#endif
+
+  // <path>/<name>[.exe]<NUL>
+  size_t totalLen = pathLen + nameLen + extLen + 1;
+  char *fullPath = (char *)malloc(totalLen);
+  char *ptr = fullPath;
+  std::memcpy(ptr, path, pathLen);
+  ptr += pathLen;
+  std::memcpy(ptr, name, nameLen);
+  ptr += nameLen;
+
+#ifdef _WIN32
+  if (extLen) {
+    std::memcpy(ptr, ".exe", 4);
+    ptr += 4;
+  }
+#endif
+
+  *ptr = 0;
+
+  return fullPath;
+}
+
+#if !defined(_WIN32) || defined(__CYGWIN__)
+void
+_swift_initRuntimePath(void *) {
```

Comment:
Could you use `SymbolInfo` here? It would eliminate the platform-specific shenanigans.

Reply:
In principle, though we'd still want to call `dyld_image_path_containing_address()` on Darwin, and `SymbolInfo` uses `DbgHelp` on Windows, right, whereas this doesn't.